Added dispense armor feature

Christian7573 opened the pull request:
I've changed two things:

1. I've added the feature reported missing in #462, the ability to equip armor to a player with a dispenser.
Note: This is WIP because upon writing this pull request I realized that I only included armor, while there are other equitable items such as heads and elytras.

2. During testing of item 1, I discovered that items with a unique dispenser behavior would only use that behavior properly "when they felt like it." (I didn't understand the dispenser slot randomization code) I rewrote `getDispenseSlot()` of GlowDispenser to fix this.

Commit:
Added the ability to eqiup armor from dispensers

Commit:
Fixed the bug where aditional dispenser behaviors would only work when they felt like it

aramperes commented on src/main/java/net/glowstone/dispenser/ArmorDispenseBehavior.java:
```diff
@@ -0,0 +1,104 @@
+
+/**
+ *
+ * @author cag45
```
We don't really use `@author` tags in Glowstone, could you remove this comment block?

Christian7573 replied:
Sure, forgot that NetBeans automatically adds those.

aramperes commented on src/main/java/net/glowstone/dispenser/ArmorDispenseBehavior.java:
```diff
@@ -0,0 +1,104 @@
+            case CHAINMAIL_CHESTPLATE: armor = Material.DIAMOND_CHESTPLATE; break;
+        }
+        
+        List<Player> players = world.getPlayers();
```
I think this functionality applies to other entities as well (those who can equip armor). You can use [`world.getNearbyEntities(location, x, y, z)`](https://hub.spigotmc.org/javadocs/spigot/org/bukkit/World.html#getNearbyEntities(org.bukkit.Location,%20double,%20double,%20double)) to look through entities in an area.

Christian7573 replied:
Ok, will implement that

aramperes commented on src/main/java/net/glowstone/block/state/GlowDispenser.java:
```diff
@@ -68,17 +89,44 @@ public boolean dispense() {
-                slot = i;
+        
+        boolean allAir = true;
+        if (inv.getItem(0).getType() != Material.AIR) allAir = false;
```
I would use a for-loop here... but can't you use `isEmpty()` in this case?

Christian7573 replied:
Didn't know that was a function, yes I will use that.

aramperes commented on src/main/java/net/glowstone/block/state/GlowDispenser.java:
```diff
@@ -68,17 +89,44 @@ public boolean dispense() {
+        if (allAir) { //Determines if the dispenser is empty, if so return slot 0
+            return 0;
+        } else {
+            
```
Extra whitespace.

Christian7573 replied:
Will be removed.

Commit:
Applied changes requested by momothereal to GlowDispenser

aramperes commented on src/main/java/net/glowstone/block/state/GlowDispenser.java:
```diff
@@ -68,17 +90,37 @@ public boolean dispense() {
+        
+        boolean isEmpty = true;
+        for (ItemStack stack : inv.getContents()) {
+            if (InventoryUtil.isEmpty(stack) == false) isEmpty = false;
```
For efficiency, you should break your loop after finding a non-empty stack. Also, use `!InventoryUtil.isEmpty(stack)`.

Christian7573 replied:
Thanks for the advice, changes made.

aramperes commented on src/main/java/net/glowstone/block/state/GlowDispenser.java:
```diff
@@ -68,17 +90,37 @@ public boolean dispense() {
+        for (ItemStack stack : inv.getContents()) {
+            if (InventoryUtil.isEmpty(stack) == false) isEmpty = false;
+        } 
+        if (isEmpty == false) {
```
Same thing here, you should use `!isEmpty`.

Christian7573 replied:
Ok

picknchew commented on src/main/java/net/glowstone/block/state/GlowDispenser.java:
```diff
@@ -68,17 +90,37 @@ public boolean dispense() {
+                
+                //Eliminates the possibility of an infinite loop, goes in order upon reaching 20 random tries
+                tries += 1;
+                if (tries == 20) {
```
Algorithm is unreliable and the number chosen is arbitrary. A better way of selecting an item would be finding all slots that have items and choosing one.

Christian7573 replied:
Is this not supposed to be arbitrary?

mastercoms replied:
Generally, you shouldn't use tries for this. An inventory isn't a black box where we don't know how many slots there are or if any of the have items in it. We can gather all slots with items through inventory methods.

picknchew commented on src/main/java/net/glowstone/dispenser/ArmorDispenseBehavior.java:
```diff
@@ -0,0 +1,95 @@
+        Player player;
+        Location location;
+        Boolean location1Test;
+        Boolean location2Test;
```
Use primitive for boolean instead. These aren't needed anyways. Just compare them directly.

Christian7573 replied:
Those were put there to keep the if statement lines short, but I'll change it if you wish.

picknchew commented on src/main/java/net/glowstone/dispenser/ArmorDispenseBehavior.java:
```diff
@@ -0,0 +1,95 @@
+        Boolean location2Test;
+        int i = 0;
+        //Loop through players to see if any are in the location where armor would be dispensed
+        while (i < players.size()) {
```
For loop is much more suitable for this

Christian7573 replied:
Agreed

picknchew commented on src/main/java/net/glowstone/dispenser/ArmorDispenseBehavior.java:
```diff
@@ -0,0 +1,95 @@
+        BlockFace facing = BlockDispenser.getFacing(block);
+        GlowBlock target = block.getRelative(facing);
+        Location location1 = target.getLocation();
+        Location location2 = location1.clone();
```
There is no need to create a new location to compare them.

Christian7573 replied:
It's not that, if a dispenser is placed at your head level it should also be able to equip items. That is why there is 2 locations, 1 for the block which the dispenser if facing and another one for 1 block below it. If you're referring to `facing` and `target` those were copy-pasted from `BucketDispenseBehavior`

picknchew replied:
There is no need to clone the objects to compare them. A simple check between the values will do.

picknchew commented on src/main/java/net/glowstone/dispenser/ArmorDispenseBehavior.java:
```diff
@@ -0,0 +1,95 @@
+    
+    @Override
+    protected ItemStack dispenseStack(GlowBlock block, ItemStack stack) {
+        Material armor = stack.clone().getType();
```
No need to clone

picknchew commented on src/main/java/net/glowstone/dispenser/ArmorDispenseBehavior.java:
```diff
@@ -0,0 +1,95 @@
+            location2Test = (location.getX() == location2.getX() && location.getY() == location2.getY() && location.getZ() == location2.getZ());
+            if (location1Test || location2Test) {
+                //Check if the aproperiate armor slot is empty, if so equip armor.
+                if (armor == Material.DIAMOND_BOOTS) {
```
Instead of doing this, it'd be much better to find a way to differentiate the different types of armor internally.

aramperes replied:
Yes, and the same system would be useful when we implement entities picking up armor on the ground (and upgrading their equipment)

Christian7573 replied:
So perhaps add an integer called Wearable to ItemType with 0 meaning cannot be worn, 1 for head, 2 for chestplate, 3 for leggings and 4 for boots?

aramperes replied:
Perhaps, or an enum could be better. Would also need a "value" level to determine which armor material is better than another, etc.

Christian7573 replied:
Okee, I'll see what I can do.

Christian7573 commented:
Ability to add armor to armor stands is not going to be implemented yet because of #471

Commit:
Prepared for armor stand compatibility
At time of commit Armor Stands are broken, so code adding armor stands
to the entity variable is broken.
* Prepared for armor stand compatibility
* Added Typeless class, most likely will be remoed in next commit.
* Minor changes to GlowDispenser

Commit:
Added ItemWearable class, but will be removed soon.
* Removed Typeless class
* Added ItemWearable class, but this will be removed in the next commit
due to conflicts with wearable items with their own classes already.
* Added armor to ItemTable

To Do:
Setup wearable item system in ItemClass

Commit:
Merged ItemWearable into ItemType
* Deleted ItemWearable
* Removed ItemWearable functionality from BlockPumpkin
* Added ItemWearable functionality to ItemType
* Added method wear to ItemType to make making items wearable easier

mastercoms commented on src/main/java/net/glowstone/block/itemtype/ItemType.java:
```diff
@@ -130,15 +130,17 @@ protected final void setMaxStackSize(int maxStackSize) {
     public void rightClickAir(GlowPlayer player, ItemStack holding) {
-        // nothing by default
+        if (fastEquip) {
+            equipStack(player, holding);
```
This will cause dupe usage, as the client sends a right click and use packet. This is not meant to check if self usage is only allowed.

Christian7573 replied:
This is for the ability to equip armor by right clicking it, as I remember checking and it not being implemented.

mastercoms replied:
See https://github.com/GlowstoneMC/Glowstone/pull/470#discussion_r115345972.

mastercoms commented on src/main/java/net/glowstone/block/itemtype/ItemType.java:
```diff
@@ -130,15 +130,17 @@ protected final void setMaxStackSize(int maxStackSize) {
      */
     public boolean canOnlyUseSelf() {
-        return false;
+        return fastEquip;
```
This is not supposed to be a getter for any fields. It's there for subclasses to define themselves as items with such behavior.

Christian7573 replied:
This is for the ability to equip armor by right clicking it, as I remember checking and it not being implemented. If it can be right clicked it should not be placed.

mastercoms replied:
See https://github.com/GlowstoneMC/Glowstone/pull/470#discussion_r115345972.

mastercoms commented on src/main/java/net/glowstone/block/itemtype/ItemType.java:
```diff
@@ -164,10 +166,67 @@ public final String toString() {
     public ItemWearablePosition getWearablePosition() {
         return wearablePosition;
     }
     public void setWearablePosition(ItemWearablePosition position) {
```
This is invalid for most items. Keep the subclass for wearable items. Is there a reason the class was removed? And what were the use cases of the class and these methods?

Christian7573 replied:
I removed it because Skull and Pumpkin had different classes that both extended from different classes, so simply extending them from ItemWearable wouldn't work

mastercoms replied:
Then use an interface with default methods for now until the new types system comes around, which will fix these inheritance issues.

Christian7573 replied:
Ok.

mastercoms commented on src/main/java/net/glowstone/block/ItemTable.java:
```diff
@@ -301,6 +299,31 @@ private void registerBuiltins() {
         reg(Material.STORAGE_MINECART, new ItemMinecart(GlowMinecart.MinecartType.CHEST));
+        
+        //Wearable Items
+        reg(Material.LEATHER_BOOTS, new ItemType().wear(1, ItemWearablePosition.FEET, true));
```
Make a subclass with a constructor for this stuff, after explaining why this functionality is needed.

mastercoms replied:
Ah, I see it's for determining the armor type of the material? You should use `EnchantmentTarget` for that, so just revert all of this unless there's more uses for `ItemWearable`.

Christian7573 replied:
Well items such as heads, pumpkins, and elytras can be work but cannot be enchanted. The integer is to determine how many armor points wearing that item gives you. 

>  Would also need a "value" level to determine which armor material is better than another.

Then for where you wear this item and then if you should be able to equip this item by right clicking it.

mastercoms replied:
Equipping armor with right click was working before. Don't include a fix for it in this PR, since it's a regression somewhere else and adding a new implementation for this feature won't fix the root issue. Though, I appreciate pointing my attention to this new bug.

Christian7573 replied:
Ok, and how did you reference a comment?

mastercoms replied:
Just posted the comment link and GitHub handles formatting.

Christian7573 committed:
Reverted changes

Reverted all changes (I think) to commit
385e67ee14995c0fc61261eb75ffd41ee7912267

Christian7573 committed:
Made Skulls Wearable

* Added ItemSkullItem, which extends ItemWearable and calls
`setPlaceAs(Material.SKULL)` in the constructor
* Removed code for right-click equip.

To Do:
* Make pumpkins wearable
* Possibly make BlockWearable class

Christian7573 committed:
Merge remote-tracking branch 'refs/remotes/GlowstoneMC/master'

Christian7573 committed:
Fixed errors

Christian7573 committed:
Merge remote-tracking branch 'refs/remotes/GlowstoneMC/master'

mastercoms commented on src/main/java/net/glowstone/block/itemtype/ItemWearablePosition.java:
```diff
@@ -0,0 +1,9 @@
+package net.glowstone.block.itemtype;
+
+public enum ItemWearablePosition {
```
This class is redundant, use EquipmentSlot.

mastercoms commented on src/main/java/net/glowstone/util/Typeless.java:
```diff
@@ -0,0 +1,12 @@
+package net.glowstone.util;
+
+public class Typeless<T> {
```
What is this for?

Christian7573 replied:
Pretty sure I deleted that, must have accidentally restored it when restoring other commits.

mastercoms commented on src/main/java/net/glowstone/block/itemtype/ItemWearable.java:
```diff
@@ -0,0 +1,60 @@
+import org.bukkit.entity.LivingEntity;
+import org.bukkit.inventory.ItemStack; 
+
+public class ItemWearable extends ItemType {
```
I still don't understand what ItemWearable is used for the dispenser logic.

If you're going to use it to equip find equipment slot type, just use `EnchantmentTarget.includes` for normal armor, and then a few extra boolean checks for the other non-enchantable materials, like pumpkins or skulls. 

You could make some sort of utility class for it too, to check for those extra materials, as well as to do the correct equipping for each equipment slot type.

I just don't think this data needs to be instanced per material in a new wearable class.

Christian7573 replied:
Oh, sorry, did not realize that EnchantmentTarget included things like ARMOR_FEET, I just saw ARMOR.

Christian7573 committed:
Moved functionality to Wearable util

* Added `net.glowstone.util.Wearable`, containing code to determine if
something is wearable and the equip method
* Removed wearable items from `ItemTable` and reverted code for
`SKULL_ITEM`
* Removed `ItemWearable` class
* Removed `ItemSkullItem` class
* Removed ItemWearable functionality from `ItemType`
* Removed `Typeless` again

Christian7573 committed:
Merge remote-tracking branch 'refs/remotes/GlowstoneMC/master'

Christian7573 committed:
Deleted `ItemWearablePosition`

Forgot to do that eariler

Christian7573 committed:
Use Wearable for ArmorDispenseBehavior

* Added `allWearables()` method to `Wearable` to get a `Set` of all
wearable items
* Used `allWearables()` in `GlowDispenser`
* Used `Wearable.equip()` in `ArmorDispenseBehavior`

Christian7573 commented:
Note: Latest commits have not been tested yet.

mastercoms commented on src/main/java/net/glowstone/util/Wearable.java:
```diff
@@ -0,0 +1,120 @@
+    
+    protected Map<Material, WearableReturn> createMap() {
+        Map<Material, WearableReturn> map2 = new HashMap<>();
+        map2.put(Material.LEATHER_BOOTS, new WearableReturn(EnchantmentTarget.ARMOR_FEET, 1));
```
Don't use `EnchantmentTarget` for specifying slot, only for determining it. Slot should be specified with `EquipmentSlot` (if needed) and it is not needed to define armor to be put in a certain slot, as that can be checked with the `includes(Material)` method in `EnchantmentTarget`. For example, `EnchantmentTarget.ARMOR_HEAD.includes(Material.GOLD_HELMET)` would return true.

Christian7573 replied:
Ok.

mastercoms commented on src/main/java/net/glowstone/util/Wearable.java:
```diff
@@ -0,0 +1,120 @@
+        public EnchantmentTarget wearPosition() {
+            return wearablePosition;
+        }
+        public int armorPoints() {
```
In Java, accessors are typically named with the prefix `get`, so for example, your method should be called `getArmorPoints()` instead of just `armorPoints()`.

mastercoms commented on src/main/java/net/glowstone/util/Wearable.java:
```diff
@@ -0,0 +1,120 @@
+        return map;
+    }
+    
+    public static void intalize() {
```
Since your map is static, you can use a static method to **initialize** your map. So, if you were to keep `createMap()` add `static` after `protected`.

Then, you can run that static method using a static block:

```java
static {
  createMap();
}
```

Christian7573 replied:
Since I didn't realize I could set map as `new HashMap<>()` before, I was defining map2 as a local variable, but whenever I tried to access it NetBeans yelled at me for "accessing a non-static variable from a static method"

mastercoms commented on src/main/java/net/glowstone/util/Wearable.java:
```diff
@@ -0,0 +1,120 @@
+    }
+    
+    protected Map<Material, WearableReturn> createMap() {
+        Map<Material, WearableReturn> map2 = new HashMap<>();
```
You can assign `map` to `new HashMap<>()` in the field declaration. It doesn't need to be separated.

Then, you can directly add entries to `map` instead of using `map2` and then assigning `map` to `map2`.

Christian7573 replied:
Oh, didn't think of that.

mastercoms commented on src/main/java/net/glowstone/util/Wearable.java:
```diff
@@ -0,0 +1,120 @@
+import java.util.HashMap;
+import java.util.Set;
+
+public final class Wearable {
```
After seeing this class, I have understood what you're trying to do. This behavior is already implemented in `GlowPlayerInventory`. Use `tryToFillSlots(armor, 36, 40)`

I have put some comments about your code in case you need to use similar Java code in the future.

Christian7573 replied:
Ok

Christian7573 commented:
So `tryToFillSlots()` from `GlowInventory` returns null if all items are placed successfully. Can I return null from `ArmorDispenseBehavior` or do I have to check if it's null and return `new ItemStack(Material.AIR)`?

mastercoms commented:
`null` is fine, but you can use `InventoryUtil.itemOrEmpty()`

Christian7573 commented:
Ok, thanks.

Christian7573 committed:
Finished ArmorDispenseBehavior

* Removed class `Wearable`
* Implemented new logic into `ArmorDispenseBehavior`
* Linked `Material.PUMPKIN` and `Material.SKULL_ITEM` with
`ArmorDispenseBehavior` in `GlowDispenser`

mastercoms commented on src/main/java/net/glowstone/block/ItemTable.java:
```diff
@@ -234,7 +234,6 @@ private void registerBuiltins() {
         reg(Material.BREWING_STAND_ITEM, new ItemPlaceAs(Material.BREWING_STAND));
         reg(Material.CAULDRON_ITEM, new ItemPlaceAs(Material.CAULDRON));
         reg(Material.FLOWER_POT_ITEM, new ItemPlaceAs(Material.FLOWER_POT));
-        reg(Material.SKULL_ITEM, new ItemPlaceAs(Material.SKULL));
```
I don't think this diff is needed.

Christian7573 replied:
What is "this diff"
If you're wondering why I removed it it's just been moved to the bottom of the list, I can't remember why though. I think it was because I used to have all wearables in one place.

mastercoms replied:
A diff is your changes from the previous revision of a file. I am just saying that it isn't necessary to move the line down.

Christian7573 replied:
Oh ok.

mastercoms replied:
Here's just a reminder to do this :)

Christian7573 replied:
Oops, forgot. Thanks for the reminder.

mastercoms commented on src/main/java/net/glowstone/block/state/GlowDispenser.java:
```diff
@@ -32,11 +33,34 @@ public static DispenseBehaviorRegistry getDispenseBehaviorRegistry() {
     public static void register() {
         // register all dispense behaviors
         DefaultDispenseBehavior bucketDispenseBehavior = new BucketDispenseBehavior();
+        ArmorDispenseBehavior armorDispenseBehavior = new ArmorDispenseBehavior();
```
Move this line to right before you register the behavior for all of the new materials. (line 42 in current file)

Christian7573 replied:
Ok

mastercoms commented on src/main/java/net/glowstone/dispenser/ArmorDispenseBehavior.java:
```diff
@@ -0,0 +1,56 @@
+        
+        //Find all nearby entities and see if they are players or armor stands
+        List<LivingEntity> entities = new ArrayList<>();
+        for (Entity cEntity : world.getNearbyEntities(location1, 3, 3, 3)) {
```
Any reason why this is named `cEntity` instead of `entity`?

Christian7573 replied:
I think I used to have a variable called entity, so I put a `c` infront of it for `currentEntity`. I'll change it to entity if you like.

mastercoms replied:
Yes please.

mastercoms commented on src/main/java/net/glowstone/dispenser/ArmorDispenseBehavior.java:
```diff
@@ -0,0 +1,56 @@
+        Material armor = stack.getType();
+        BlockFace facing = BlockDispenser.getFacing(block);
+        GlowBlock target = block.getRelative(facing);
+        Location location1 = target.getLocation();
```
Could you explain to me what these locations are used for?

Christian7573 replied:
`Location1` is the absolute location of the block the dispenser if facing. `Location2` is one block below that, because in my vanilla tests dispensers can equip armor to you if they are at your head level.

mastercoms replied:
You can just compare
entity.getLocation() and entity.getEyeLocation() to the target location.

Christian7573 replied:
Is `getEyeLocation()` a block location?

mastercoms replied:
It is a precise location so you'll have to compare the block coordinates.

Christian7573 replied:
Ok

mastercoms commented on src/main/java/net/glowstone/dispenser/ArmorDispenseBehavior.java:
```diff
@@ -0,0 +1,56 @@
+        }  
+        
+        Location location;
+        Boolean location1Test;
```
Use the `boolean` primitive type.

Christian7573 replied:
Ok

Christian7573 commented:
For whatever reason, GitHub isn't letting me respond to your comment about tries. First of all, I assumed from the previous code for getting a slot that it was a random decision, if it's not let me know and I'll make it go in order. If it is, I'm hearing that you want to randomly select from the only the ones that have items in them, which makes sense.

mastercoms commented:
Yes, when there is more than one slot occupied by items in a dispenser, a random occupied slot is dispensed.

Christian7573 committed:
Made Changes Requested

Christian7573 committed:
Merge remote-tracking branch 'refs/remotes/GlowstoneMC/master'

Christian7573 committed:
Minor Changes

* Changed `cEntity` to `entity` in `ArmorDispenseBehavior`
* Changed `GlowDispenser` to use `InventoryUtil.getRandomSlot()`
* Changed `(notEmpty.size() == 0)` to `(notEmpty.isEmpty())` in
`InventoryUtil.getRandomSlot()`

mastercoms commented on src/main/java/net/glowstone/block/state/GlowDispenser.java:
```diff
@@ -68,17 +94,7 @@ public boolean dispense() {
 
     public int getDispenseSlot() {
         Inventory inv = getInventory();
```
I don't think this is needed. Reference `getInventory()` directly in the `getRandomSlot()` call.

Christian7573 replied:
Nice idea

Christian7573 committed:
Removed unused imports

mastercoms commented on src/main/java/net/glowstone/dispenser/ArmorDispenseBehavior.java:
```diff
@@ -0,0 +1,54 @@
+        Material armor = stack.getType();
+        BlockFace facing = BlockDispenser.getFacing(block);
+        GlowBlock target = block.getRelative(facing);
+        Location location1 = target.getLocation();
```
Rename this to `targetLocation`, please.

Christian7573 replied:
Sure

mastercoms commented on src/main/java/net/glowstone/dispenser/ArmorDispenseBehavior.java:
```diff
@@ -0,0 +1,54 @@
+        
+        //Find all nearby entities and see if they are players or armor stands
+        List<LivingEntity> entities = new ArrayList<>();
+        for (Entity entity : world.getNearbyEntities(location1, 3, 3, 3)) {
```
Inline `targetLocation.getWorld()`.

Christian7573 replied:
Ok.

mastercoms commented on src/main/java/net/glowstone/dispenser/ArmorDispenseBehavior.java:
```diff
@@ -0,0 +1,54 @@
+        boolean location2Test;
+        //Loop through entities to see if any are in the location where armor would be dispensed
+        for (LivingEntity player : entities) {          
+            location = player.getLocation().clone();
```
I don't think this is needed. Just use the block coordinate getters.

Christian7573 replied:
By "block coordinate getters" are you referring to `location.getBlockX()` and Y and Z?

mastercoms commented on src/main/java/net/glowstone/dispenser/ArmorDispenseBehavior.java:
```diff
@@ -0,0 +1,54 @@
+            location.setY(location.getBlockY());
+            location.setZ(location.getBlockZ());            
+            location1Test = (location.getX() == location1.getX() && location.getY() == location1.getY() && location.getZ() == location1.getZ());
+            location2Test = (location.getX() == player.getEyeLocation().getBlockX() && player.getEyeLocation().getBlockY() == location1.getY() && player.getEyeLocation().getBlockZ() == location1.getZ());
```
Compare the target's location to the player's eye location. Right now, the code is comparing the player's location to their eye location.

Christian7573 replied:
Oh wow can't believe I did that. Will be fixed.

mastercoms commented on src/main/java/net/glowstone/dispenser/ArmorDispenseBehavior.java:
```diff
@@ -0,0 +1,54 @@
+                return ((GlowInventory) ((Player) player).getInventory()).tryToFillSlots(stack, 36, 40);
+            }
+        }
+        return defaultBehavior.dispense(block, stack); //Fallback
```
Comments should have a space after the `//`

Christian7573 replied:
Ok

Christian7573 committed:
Made Requested Changes

* Removed unused `armor` variable in `ArmorDispenseBehavior`
* Removed unused imports in `ArmorDispenseBehavior`
* Added spaces in comments
* Removed `getDispenseSlot()` from `GlowDispenser`
* Changed code for slot selection in `GlowDispenser`
* Removed `world` variable from `GlowDispenser`
* Changed a few variable names in `GlowDispenser`
* Fixed error in location checking code in `GlowDispenser`

Christian7573 committed:
Merge remote-tracking branch 'refs/remotes/GlowstoneMC/master'

mastercoms commented on src/main/java/net/glowstone/block/state/GlowDispenser.java:
```diff
@@ -66,21 +90,6 @@ public boolean dispense() {
         return true;
     }
 
-    public int getDispenseSlot() {
```
I meant
```java
public int getDispenseSlot() {
  return InventoryUtil.getRandomSlot(random, getInventory(), true);
}
```

Christian7573 replied:
Oh ok.

mastercoms commented on src/main/java/net/glowstone/block/state/GlowDispenser.java:
```diff
@@ -32,11 +33,34 @@ public static DispenseBehaviorRegistry getDispenseBehaviorRegistry() {
         getDispenseBehaviorRegistry().putBehavior(Material.LAVA_BUCKET, bucketDispenseBehavior);
         getDispenseBehaviorRegistry().putBehavior(Material.BUCKET, new EmptyBucketDispenseBehavior());
         getDispenseBehaviorRegistry().putBehavior(Material.FLINT_AND_STEEL, new FlintAndSteelDispenseBehavior());
         getDispenseBehaviorRegistry().putBehavior(Material.TNT, new TNTDispenseBehavior());
```
Move `ArmorDispenseBehavior armorDispenseBehavior = new ArmorDispenseBehavior();` to here

mastercoms commented on src/main/java/net/glowstone/dispenser/ArmorDispenseBehavior.java:
```diff
@@ -0,0 +1,50 @@
+            location.setX(location.getBlockX());
+            location.setY(location.getBlockY());
+            location.setZ(location.getBlockZ());            
+            targetLocationTest1 = (player.getLocation().getBlockX() == targetLocation.getX() && player.getLocation().getBlockY() == targetLocation.getY() && player.getLocation().getBlockZ() == targetLocation.getZ());
```
Make sure you're using getBlockX, getBlockY and getBlockZ for all locations.

Christian7573 replied:
Ok

Christian7573 replied:
Wait, since `targetLocation` is already a block location do I need to use it there too?

mastercoms replied:
Oh whoops, you're right. I wasn't thinking about that. Simple getX and similar are fine for locations of blocks.

mastercoms commented on src/main/java/net/glowstone/dispenser/ArmorDispenseBehavior.java:
```diff
@@ -0,0 +1,50 @@
+            }
+        }  
+        
+        Location location;
```
This location variable isn't needed anymore since you're accessing the player location directly.

Christian7573 replied:
Will be removed

mastercoms commented on src/main/java/net/glowstone/dispenser/ArmorDispenseBehavior.java:
```diff
@@ -0,0 +1,50 @@
+    @Override
+    protected ItemStack dispenseStack(GlowBlock block, ItemStack stack) {
+        BlockFace facing = BlockDispenser.getFacing(block);
+        GlowBlock target = block.getRelative(facing);
```
It is a bit expensive to get a block, so try cloning `block`'s location and then adding facing.getModX,Y and Z to that location to get `targetLocation`.

Christian7573 replied:
Sure

Christian7573 committed:
Made changes requested

* Added `getDispenseSlot()` back to `GlowDispenser`
* Moved SKULL_ITEM in `ItemTable` back to it's origional location
* Removed unused variable `location` from `ArmorDispenseBehavior`
* Removed variable `target` from `ArmorDispenseBehavior` and changed
code to use `BlockFace.getModX()`, Y and Z

mastercoms commented on src/main/java/net/glowstone/block/state/GlowDispenser.java:
```diff
@@ -57,30 +82,19 @@ public boolean dispense() {
             return false;
         }
-
+        
```
Extra whitespace

mastercoms commented on src/main/java/net/glowstone/block/state/GlowDispenser.java:
```diff
@@ -57,30 +82,19 @@ public boolean dispense() {
         return true;
     }
-
+    
```
Extra whitespace

mastercoms commented on src/main/java/net/glowstone/block/state/GlowDispenser.java:
```diff
@@ -57,30 +82,19 @@ public boolean dispense() {
+        return InventoryUtil.getRandomSlot(random, getInventory(), true);
     }
-
+    
```
Extra whitespace

mastercoms commented:
Hi, this looks good to go, there's just some extra whitespace, but that's no big deal. I can clean that up, just remember to clean up your whitespace next time please.

Christian7573 commented:
Ok

Christian7573 commented:
Ok

mastercoms commented:
Thank you for your contribution!
